build/generate-licenses.go

```diff
@@ -115,8 +125,156 @@ func main() {
+				continue
+			}
+			if md5 != "" {
+				_, ok := sameFiles[md5]
+				if !ok {
+					sameFiles[md5] = make([]string, 0)
+				}
+				if !contains(sameFiles[md5], preLicenseName) {
+					sameFiles[md5] = append(sameFiles[md5], preLicenseName)
+				}
+				sameFiles[md5] = append(sameFiles[md5], licenseName)
+			}
+			preFile = out
+			preLicenseName = licenseName
 		}
 	}
 
+	// generate convert license name map
+	sameLicenses := make(map[string]string)
+	for _, fileNames := range sameFiles {
+		key := getLicenseKey(fileNames)
+		for _, fileName := range fileNames {
+			sameLicenses[fileName] = key
+		}
+	}
+	// save convert license name map to file
+	bytes, err := json.Marshal(sameLicenses)
+	if err != nil {
+		log.Fatalf("Failed to create json bytes. %s", err)
+		return
+	}
+	out, err := os.Create(sameLicensesDestination)
+	if err != nil {
+		log.Fatalf("Failed to create new file. %s", err)
+	}
+	defer out.Close()
+	_, err = out.Write(bytes)
+	if err != nil {
+		log.Fatalf("Failed to write same licenses json file. %s", err)
+	}
+
 	fmt.Println("Done")
 }
+
+// getSameFileMD5 returns md5 of the input file, if the content of input files are same
+func getSameFileMD5(f1, f2 *os.File) (string, error) {
+	if f1 == nil || f2 == nil {
+		return "", nil
+	}
+
+	// check file size
+	fs1, err := f1.Stat()
+	if err != nil {
+		return "", err
+	}
+	fs2, err := f2.Stat()
+	if err != nil {
+		return "", err
+	}
+
+	if fs1.Size() != fs2.Size() {
+		return "", nil
+	}
+
+	// check content
+	chunkSize := 1024
+	_, err = f1.Seek(0, 0)
+	if err != nil {
+		return "", err
+	}
+	_, err = f2.Seek(0, 0)
+	if err != nil {
+		return "", err
+	}
+
+	var totalBytes []byte
+	for {
+		b1 := make([]byte, chunkSize)
+		_, err1 := f1.Read(b1)
+
+		b2 := make([]byte, chunkSize)
+		_, err2 := f2.Read(b2)
+
+		totalBytes = append(totalBytes, b1...)
+
+		if err1 != nil || err2 != nil {
+			if err1 == io.EOF && err2 == io.EOF {
+				md5 := md5.Sum(totalBytes)
+				return string(md5[:]), nil
+			} else if err1 == io.EOF || err2 == io.EOF {
+				return "", nil
+			} else if err1 != nil {
+				return "", err1
+			} else if err2 != nil {
+				return "", err2
+			}
+		}
+
+		if !bytes.Equal(b1, b2) {
+			return "", nil
+		}
+	}
+}
+
+func getLicenseKey(fnl []string) string {
+	if len(fnl) == 0 {
+		return ""
+	}
+
+	shortestItem := func(list []string) string {
+		s := list[0]
+		for _, l := range list[1:] {
+			if len(l) < len(s) {
+				s = l
+			}
+		}
+		return s
+	}
+	allHasPrefix := func(list []string, s string) bool {
+		for _, l := range list {
+			if !strings.HasPrefix(l, s) {
+				return false
+			}
+		}
+		return true
+	}
+
+	sl := shortestItem(fnl)
+	slv := strings.Split(sl, "-")
+	var result string
+	for i := len(slv); i >= 0; i-- {
+		result = strings.Join(slv[:i], "-")
+		if allHasPrefix(fnl, result) {
+			return result
+		}
+	}
+	return ""
+}
+
+func contains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Couldn't this be replaced by `slices.Contains`? https://pkg.go.dev/slices#Contains

Reply:
Done in https://github.com/go-gitea/gitea/pull/24872/commits/d6ba4af183a94f2ae3c28e11fa094a34a96b988a.